lib/datadog/tracing/distributed/datadog.rb

```diff
@@ -0,0 +1,150 @@
+        #
+        # DEV: This method accesses global state (the active trace) to record its error state as a trace tag.
+        # DEV: This means errors cannot be reported if there's not active span.
+        # DEV: Ideally, we'd have a dedicated error reporting stream for all of ddtrace.
+        # DEV: The same comment applies to the {.extract_tags}.
+        def inject_tags(digest, data)
+          return if digest.trace_distributed_tags.nil? || digest.trace_distributed_tags.empty?
+
+          if ::Datadog.configuration.tracing.x_datadog_tags_max_length <= 0
+            active_trace = Tracing.active_trace
+            active_trace.set_tag('_dd.propagation_error', 'disabled') if active_trace
+            return
+          end
+
+          encoded_tags = DatadogTagsCodec.encode(digest.trace_distributed_tags)
+
+          if encoded_tags.size > ::Datadog.configuration.tracing.x_datadog_tags_max_length
+            active_trace = Tracing.active_trace
+            active_trace.set_tag('_dd.propagation_error', 'inject_max_size') if active_trace
+
+            ::Datadog.logger.warn(
+              "Failed to inject x-datadog-tags: tags are too large (size:#{encoded_tags.size} " \
+                "limit:#{::Datadog.configuration.tracing.x_datadog_tags_max_length}). This limit can be configured " \
+                'through the DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH environment variable.'
+            )
+            return
+          end
+
+          data[@tags] = encoded_tags
+        rescue => e
+          active_trace = Tracing.active_trace
+          active_trace.set_tag('_dd.propagation_error', 'encoding_error') if active_trace
+          ::Datadog.logger.warn(
+            "Failed to inject x-datadog-tags: #{e.class.name} #{e.message} at #{Array(e.backtrace).first}"
+          )
+        end
+
+        # Import `x-datadog-tags` tags as trace distributed tags.
+        # Only tags that have the `_dd.p.` prefix are processed.
+        def extract_tags(data)
+          tags = data[@tags]
+          return if !tags || tags.empty?
+
+          if ::Datadog.configuration.tracing.x_datadog_tags_max_length <= 0
+            active_trace = Tracing.active_trace
+            active_trace.set_tag('_dd.propagation_error', 'disabled') if active_trace
+            return
+          end
+
+          if tags.size > ::Datadog.configuration.tracing.x_datadog_tags_max_length
+            active_trace = Tracing.active_trace
+            active_trace.set_tag('_dd.propagation_error', 'extract_max_size') if active_trace
+
+            ::Datadog.logger.warn(
+              "Failed to extract x-datadog-tags: tags are too large (size:#{tags.size} " \
+                "limit:#{::Datadog.configuration.tracing.x_datadog_tags_max_length}). This limit can be configured " \
+                'through the DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH environment variable.'
+            )
+            return
+          end
```

Comment:
This seems to be almost exactly the same as for `#inject_tags` (I believe the only change is `inject_max_size` vs `extract_max_size` for the kind of error). Perhaps consider extracting to a shared method?

Reply:
Regarding sharing code between `extract_tags` and `inject_tags`, I don't think I can refactor it to share code.
Every other line has a different error message or string.

Unless you are referring to a specific line range, I don't have good ideas to refactor this.

Reply:
I was thinking along the lines of:

```diff
diff --git a/lib/datadog/tracing/distributed/datadog.rb b/lib/datadog/tracing/distributed/datadog.rb
index ffaebbf83..fa1d21b39 100644
--- a/lib/datadog/tracing/distributed/datadog.rb
+++ b/lib/datadog/tracing/distributed/datadog.rb
@@ -72,25 +72,12 @@ module Datadog
         # DEV: Ideally, we'd have a dedicated error reporting stream for all of ddtrace.
         def inject_tags(digest, data)
           return if digest.trace_distributed_tags.nil? || digest.trace_distributed_tags.empty?
-
-          if ::Datadog.configuration.tracing.x_datadog_tags_max_length <= 0
-            active_trace = Tracing.active_trace
-            active_trace.set_tag('_dd.propagation_error', 'disabled') if active_trace
-            return
-          end
+          return try_setting_propagation_error_tag_on_active_trace(error_kind: 'disabled') if datadog_tags_disabled?

           encoded_tags = DatadogTagsCodec.encode(digest.trace_distributed_tags)

-          if encoded_tags.size > ::Datadog.configuration.tracing.x_datadog_tags_max_length
-            active_trace = Tracing.active_trace
-            active_trace.set_tag('_dd.propagation_error', 'inject_max_size') if active_trace
-
-            ::Datadog.logger.warn(
-              'Failed to inject x-datadog distributed tracing tags: too many tags for configured limit ' \
-              "(size:#{encoded_tags.size} >= limit:#{::Datadog.configuration.tracing.x_datadog_tags_max_length}). This " \
-              'limit can be configured with the DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH environment variable.'
-            )
-            return
+          if too_many_tags?(encoded_tags.size, situation: 'inject')
+            return try_setting_propagation_error_tag_on_active_trace(error_kind: 'inject_max_size')
           end

           data[@tags_key] = encoded_tags
@@ -111,23 +98,10 @@ module Datadog
         def extract_tags(fetcher)
           tags = fetcher[@tags_key]
           return if !tags || tags.empty?
+          return try_setting_propagation_error_tag_on_active_trace(error_kind: 'disabled') if datadog_tags_disabled?

-          if ::Datadog.configuration.tracing.x_datadog_tags_max_length <= 0
-            active_trace = Tracing.active_trace
-            active_trace.set_tag('_dd.propagation_error', 'disabled') if active_trace
-            return
-          end
-
-          if tags.size > ::Datadog.configuration.tracing.x_datadog_tags_max_length
-            active_trace = Tracing.active_trace
-            active_trace.set_tag('_dd.propagation_error', 'extract_max_size') if active_trace
-
-            ::Datadog.logger.warn(
-              "Failed to extract x-datadog-tags: tags are too large (size:#{tags.size} >= " \
-                "limit:#{::Datadog.configuration.tracing.x_datadog_tags_max_length}). This limit can be configured " \
-                'through the DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH environment variable.'
-            )
-            return
+          if too_many_tags?(tags.size, situation: 'extract')
+            return try_setting_propagation_error_tag_on_active_trace(error_kind: 'inject_max_size')
           end

           tags_hash = DatadogTagsCodec.decode(tags)
@@ -144,6 +118,28 @@ module Datadog
           )
         end

+        def try_setting_propagation_error_tag_on_active_trace(error_kind:)
+          active_trace = Tracing.active_trace
+          active_trace.set_tag('_dd.propagation_error', error_kind) if active_trace
+          nil
+        end
+
+        def datadog_tags_disabled?
+          ::Datadog.configuration.tracing.x_datadog_tags_max_length <= 0
+        end
+
+        def too_many_tags?(tags_size, situation:)
+          return false if tags_size <= ::Datadog.configuration.tracing.x_datadog_tags_max_length
+
+          ::Datadog.logger.warn(
+            "Failed to #{situation} x-datadog-tags: too many tags for configured limit (size:#{tags_size} >= " \
+              "limit:#{::Datadog.configuration.tracing.x_datadog_tags_max_length}). This limit can be configured " \
+              'through the DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH environment variable.'
+          )
+
+          true
+        end
+
         # We want to exclude tags that we don't want to propagate downstream.
         EXCLUDED_TAG = '_dd.p.upstream_services'
         private_constant :EXCLUDED_TAG
```

Reply:
Sweet patch, applied! Looks great after, thank you so much!